<a href="https://colab.research.google.com/github/Lord-Kanzler/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment_ALEX_KAISER.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
## YOUR CODE HERE ##

In [6]:
#Loading Data
import pandas as pd
import numpy as np
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

df = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])

df.head()

--2020-02-17 20:08:49--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-02-17 20:08:50 (278 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [7]:
df.dtypes

party                   object
handicapped-infants     object
water-project           object
budget                  object
physician-fee-freeze    object
el-salvador-aid         object
religious-groups        object
anti-satellite-ban      object
aid-to-contras          object
mx-missile              object
immigration             object
synfuels                object
education               object
right-to-sue            object
crime                   object
duty-free               object
south-africa            object
dtype: object

In [8]:
# replacing values. "y": 1, "n": 0, "?": np.NaN
df = df.replace({"y": 1, "n": 0, "?": np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [9]:
# Cleaning data
dem = df[df['party']=='democrat']

dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [10]:
rep = df[df['party']=='republican']

rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [11]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [12]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [0]:
# finding differences. Testing all coluymns for practice.

In [0]:
# import stats
from scipy import stats 

In [16]:
# handicapped-infants
stats.ttest_ind(rep['handicapped-infants'], dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [18]:
# water-project	
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [19]:
# budget
stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

In [20]:
# physician-fee-freeze
stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

In [21]:
# el-salvador-aid
stats.ttest_ind(rep['el-salvador-aid'], dem['el-salvador-aid'], nan_policy='omit')

Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)

In [22]:
# religious-groups
stats.ttest_ind(rep['religious-groups'], dem['religious-groups'], nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

In [23]:
# anti-satellite-ban
stats.ttest_ind(rep['anti-satellite-ban'], dem['anti-satellite-ban'], nan_policy='omit')

Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)

In [25]:
# aid-to-contras
stats.ttest_ind(rep['aid-to-contras'], dem['aid-to-contras'], nan_policy='omit')

Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

In [26]:
# mx-missile
stats.ttest_ind(rep['mx-missile'], dem['mx-missile'], nan_policy='omit')

Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)

In [27]:
# immigration	
stats.ttest_ind(rep['immigration'], dem['immigration'], nan_policy='omit')

Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066)

In [30]:
# right-to-sue
stats.ttest_ind(rep['right-to-sue'], dem['right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

In [28]:
# synfuels
stats.ttest_ind(rep['synfuels'], dem['synfuels'], nan_policy='omit')

Ttest_indResult(statistic=-8.293603989407588, pvalue=1.5759322301054064e-15)

In [29]:
# education
stats.ttest_ind(rep['education'], dem['education'], nan_policy='omit')

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

In [31]:
# crime
stats.ttest_ind(rep['crime'], dem['crime'], nan_policy='omit')

Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

In [32]:
# duty-free
stats.ttest_ind(rep['duty-free'], dem['duty-free'], nan_policy='omit')

Ttest_indResult(statistic=-12.853146132542978, pvalue=5.997697174347365e-32)

In [33]:
# south-africa
stats.ttest_ind(rep['south-africa'], dem['south-africa'], nan_policy='omit')

Ttest_indResult(statistic=-6.849454815841208, pvalue=3.652674361672226e-11)

**Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01**


In [0]:
### south-africa

$ H_0: \bar{x}_{r} == \bar{x}_{d}$
Democrat support for the 'south-africa' bill is equal to republican support for the 'south-africa' bill.


$ H_0: \bar{x}_{r} \neq \bar{x}_{d}$
Democrat support for the 'south-africa' bill is not equal to republican support for the 'south-africa' bill.

Confidence Level: 99%

In [34]:
stats.ttest_ind(dem['south-africa'], rep['south-africa'], nan_policy='omit')

Ttest_indResult(statistic=6.849454815841208, pvalue=3.652674361672226e-11)

Conclusion:

Based on a t-statistic of ~6.85 and a p-value of $3.65*10^{-11}$. I reject the null hypothesis that democrats and republicans support the 'south-africa' bill at the same level/rate. And I suggest the alternative that democrat and republican levels of support for this bill are different. 

**Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01**

In [0]:
### physician-fee-freeze

$ H_0: \bar{x}_{r} == \bar{x}_{d}$
Republican support for the 'physician-fee-freeze' bill is equal to democrat support for the 'physician-fee-freeze' bill.


$ H_0: \bar{x}_{r} \neq \bar{x}_{d}$
Republican support for the 'physician-fee-freeze' bill is not equal to democrat support for the 'physician-fee-freeze' bill.


In [35]:
stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

Conclusion:

Based on a t-statistic of ~49.37 and a p-value of $1.99*10^{-177}$. I reject the null hypothesis that republicans and democrates support the 'physician-fee-freeze' bill at the same level/rate. And I suggest the alternative that republican and democrat levels of support for this bill are different. 

**Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)**

In [0]:
### water-project

$ H_0: \bar{x}_{r} == \bar{x}_{d}$
Republican support for the 'water-project' bill is equal to democrat support for the 'water-project' bill.


$ H_0: \bar{x}_{r} \neq \bar{x}_{d}$
Republican support for the 'water-project' bill is not equal to democrat support for the 'water-project' bill.


In [36]:
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

Conclusion:

Based on a t-statistic of ~0.089 and a p-value of 0.93, I fail to reject the null hypothesis that republicans and democrates support the 'water-project' bill at the same level/rate. 

**Stretch goals:**

In [0]:
#Function

def two_sample_Ttest(var_a, var_b):
  x = stats.ttest_ind(var_a, var_b, nan_policy='omit')
  return x

In [40]:
two_sample_Ttest(rep['physician-fee-freeze'],dem['physician-fee-freeze'])

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

In [0]:
t_value, p_value = ttest_ind(sample_one, 
                             sample_two, 
                     nan_policy='omit')

In [0]:
## Hypothesis testing to personal data

Since we don't have a personal project yet, I just use a old example I covered in a class previously.

In [0]:
# Using data from The Statistical Sleuth. link do data set: https://github.com/Lord-Kanzler/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Tuna.csv

Average mercury level in albacore tuna: µAlbacore 

Average mercury level in yellowfin tuna: µYellowfin 

H0: 
µYellowfin= µAlbacore 
 The average mercury levels in yellow fin tuna and albacore tuna do not differ. 

Ha: 
µYellowfin≠ µAlbacore
The average mercury levels in yellow fin tuna and albacore tuna are different. 

In [57]:
df = pd.read_csv('https://raw.githubusercontent.com/Lord-Kanzler/DS-Unit-1-Sprint-2-Statistics/master/module1/Tuna.csv')
df.head(50)

Tuna  Mercury
0   albacore    0.0000
1   albacore    0.4100
2   albacore    0.8200
3   albacore    0.3200
4   albacore    0.0360
5   albacore    0.2800
6   albacore    0.2900
7   albacore    0.3400
8   albacore    0.3600
9   albacore    0.4200
10  albacore    0.5100
11  albacore    0.4800
12  albacore    0.4010
13  albacore    0.3000
14  albacore    0.3000
15  albacore    0.3700
16  albacore    0.3240
17  albacore    0.5690
18  albacore    0.3850
19  albacore    0.2700
20  albacore    0.3760
21  albacore    0.3500
22  albacore    0.3000
23  albacore    0.3240
24  albacore    0.3700
25  albacore    0.5690
26  albacore    0.3500
27  albacore    0.3500
28  albacore    0.3850
29  albacore    0.3850
30  albacore    0.2700
31  albacore    0.3760
32  albacore    0.3760
33  albacore    0.2700
34  albacore    0.3000
35  albacore    0.2720
36  albacore    0.3970
37  albacore    0.4750
38  albacore    0.3900
39  albacore    0.5500
40  albacore    0.0409
41  albacore    0.3210
42  albacore    0.3960
43  yellowfin   0.0000
44  yellowfin   0.0000
45  yellowfin   0.0600
46  yellowfin   0.0900
47  yellowfin   0.1700
48  yellowfin   0.2100
49  yellowfin   0.2100

In [62]:
#Cleaning data
albacore = df[df['Tuna']=='albacore ']

albacore.head(10)

Tuna  Mercury
0  albacore     0.000
1  albacore     0.410
2  albacore     0.820
3  albacore     0.320
4  albacore     0.036
5  albacore     0.280
6  albacore     0.290
7  albacore     0.340
8  albacore     0.360
9  albacore     0.420

In [60]:
#Cleaning data
yellowfin = df[df['Tuna']=='yellowfin']

yellowfin.head(10)

Tuna  Mercury
43  yellowfin     0.00
44  yellowfin     0.00
45  yellowfin     0.06
46  yellowfin     0.09
47  yellowfin     0.17
48  yellowfin     0.21
49  yellowfin     0.21
50  yellowfin     0.22
51  yellowfin     0.23
52  yellowfin     0.24

In [63]:
#Running  two-sample t-test using "two_sample_Ttest" function from above.
two_sample_Ttest(albacore['Mercury'],yellowfin['Mercury'])

Ttest_indResult(statistic=0.08749593775826107, pvalue=0.9303416936890726)

Conclusion:

There is not much evidence against the null hypothesis H0 (t-statisitc = 0.087, p-value = 0.930)

There is not much evidence suggesting differing average mercury levels between yellow fin tuna and albacore tuna. 